poc/vdaf_poplar1.sage

```diff
@@ -11,198 +11,206 @@ from sagelib.vdaf import Vdaf, test_vdaf
                   nonce, public_share, input_share):
         dst = VERSION + I2OSP(Poplar1.ID, 4)
         (level, prefixes) = agg_param
-        (key, corr_seed, corr_inner, corr_leaf) = \
-            Poplar1.decode_input_share(input_share)
+        (idpf_public_share, corr_inner, corr_leaf) = \
+            Poplar1.decode_public_share(public_share)
+        (key, corr_seed) = Poplar1.decode_input_share(input_share)
+
+        Field = Poplar1.Idpf.current_value_type(level)[0]
 
         # Evaluate the IDPF key at the given set of prefixes.
-        value = Poplar1.Idpf.eval(
-            agg_id, public_share, key, level, prefixes)
-
-        # Get correlation shares for the given level of the IDPF tree.
-        #
-        # Implementation note: Computing the shares of `(a, b, c)`
-        # requires expanding PRG seeds into a vector of field elements
-        # of length proportional to the level of the tree. Typically
-        # the IDPF will be evaluated incrementally beginning with
-        # `level == 0`. Implementations can save computation by
-        # storing the intermediate PRG state between evaluations.
-        corr_prg = Poplar1.Idpf.Prg(corr_seed, dst + byte(agg_id))
-        for current_level in range(level+1):
-            Field = Poplar1.Idpf.current_field(current_level)
-            (a_share, b_share, c_share) = corr_prg.next_vec(Field, 3)
-        (A_share, B_share) = corr_inner[2*level:2*(level+1)] \
-            if level < Poplar1.Idpf.BITS - 1 else corr_leaf
-
-        # Compute the Aggregator's first round of the sketch. These are
-        # called the "masked input values" [BBCGGI21, Appendix C.4].
-        Field = Poplar1.Idpf.current_field(level)
-        verify_rand_prg = Poplar1.Idpf.Prg(verify_key,
-            dst + Poplar1.verify_context(nonce, level, prefixes))
-        verify_rand = verify_rand_prg.next_vec(Field, len(prefixes))
-        sketch_share = [a_share, b_share, c_share]
+        value = Poplar1.Idpf.eval(agg_id,
+                                  idpf_public_share,
+                                  key,
+                                  level,
+                                  prefixes)
+
+        # Leader expands randomness
+        if agg_id == 0:
+            prg = Poplar1.Idpf.Prg(corr_seed, dst + byte(255))
+            auth = prg.next_vec(Poplar1.Idpf.FieldInner,
+                                Poplar1.Idpf.BITS - 1)
+            auth += prg.next_vec(Poplar1.Idpf.FieldLeaf, 1)
+
+        # Prepare one-hotness and boundnedness verification
+        # and output share
+        # sum over all prefixes for boundedness verification
         out_share = []
-        for (i, r) in enumerate(verify_rand):
-            (data_share, auth_share) = value[i]
-            sketch_share[0] += data_share * r
-            sketch_share[1] += data_share * r^2
-            sketch_share[2] += auth_share * r
+        data_sum = Field(0)
+        auth_sum = Field(0)
```

Comment:
Is the Field type is not necessarily the same for data and auth?

Reply:
It should be the same type because auth is supposed to equal delta * (data) and this operation is only defined if they're the same type.